test/parallel/test-stream-pipeline.js

```diff
@@ -13,6 +13,15 @@ const assert = require('assert');
 const http = require('http');
 const { promisify } = require('util');
 
+function mustCall(fn) {
+  if (!fn) {
+    return common.mustCall();
+  }
+  return common.mustCall((...args) => {
+    new Promise(() => fn(...args));
+  });
+}
```

Comment:
@BridgeAR This is **very** strange. Unless wrapping inside a `Promise` exceptions inside `common.mustCall` will not cause failure... what's going on?

Reply:
@benjamingr: Who else is a good ping on this kind of issue?

Reply:
I don't understand this change, can you please articulate?

Reply:
We have assertion that are **currently** failing. However, the tests are passing anyway since the assertions are just simply swallowed. I have no idea why. But changing this makes the test properly fail.

Reply:
@nodejs/testing

Reply:
@ronag The line that's failing is https://github.com/nodejs/node/pull/31835/files#diff-90ee5bb811a72e51f73823660cd1665bR758. I'm guessing that the problem is related to the use of `await` a few lines above.

Reply:
Should passing an `async` function to `pipeline()` be expected to work? I would imagine that the error events and whatnot might not be compatible with promises?

Reply:
@ronag could you provide a standalone reproduce for this one?
I tried (based on 'exceptions inside common.mustCall will not cause failure')
```js
common.mustCall(() => {
  throw new Error('hello');
})();
```
and
```js
const fun = common.mustCall(() => {
  throw new Error('hello');
});
Promise.resolve(42).then(fun);
```
and they worked just fine.

Reply:
If you checkout master and run the test: `test-stream-pipeline.js`

It should fail at e.g. https://github.com/nodejs/node/blob/master/test/parallel/test-stream-pipeline.js#L777 but doesn't.

Maybe something in pipeline is swallowing the error. I need to dig into that.

Reply:
I think this might have to do with exceptions inside `Promise.catch` does not call unhandledRejection.

I found a workaround.